src/ad-bidders/prebid/utils/id-retriever.ts

```diff
@@ -1,12 +1,33 @@
 import { communicationService, eventsRepository } from '@ad-engine/communication';
-import { pbjsFactory, utils } from '@ad-engine/core';
+import { pbjsFactory, targetingService, utils } from '@ad-engine/core';
 
 export class IdRetriever {
 	private static _instance: IdRetriever;
-
+	private readonly ID_STRING_LENGTH = 16;
 	private readonly logGroup = 'bidder-id-retriever';
 	private pbjs: Pbjs;
 
+	private readonly ID_MAP = {
+		0: 'yahoo.com',
+		1: 'id5-sync.com',
+		2: 'liveConnect',
+		3: 'Experian',
+		4: 'liveintent.com',
+	};
+
+	// Generated IDString will be sent to GAM in this form:
+	// PAAAAxxxxxxxxxxxxxxxx
+	// Where P is Present, A is Absent and x is empty placeholder for new IDs
+	async generateBoiString() {
+		const userIds: string[] = (await this.getIds()).map((id) => id.source);
+		const idString = new Array(this.ID_STRING_LENGTH).fill('x');
```

Comment:
```suggestion
		const idString = 'x'.repeat(this.ID_STRING_LENGTH);
```
no need to create the Array, we can use `String.prototype.repeat` and skip the final `join('')`
[MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/repeat)

Reply:
Tbh I'm not convinced - it's much easier to replace a single index in Array than in String. In case we want to use strings in there, we need to use the substring function to modify it. Meanwhile - we can just access a particular object in Array to modify it, and then at the end of this function, we can convert it to a string. Using idString[0] we can only retrieve value from particular index but NOT modify it.